ddtrace/propagation/http.py

```diff
@@ -483,10 +492,214 @@ def _extract(headers):
+
+        if len(trace_id_hex) == 32 and len(span_id_hex) == 16 and len(trace_flags_hex) >= 2:
+            trace_id = _hex_id_to_dd_id(trace_id_hex)
+            span_id = _hex_id_to_dd_id(span_id_hex)
+
+            # All 0s are invalid values
+            assert trace_id != 0
+            assert span_id != 0
+
+            trace_flags = _hex_id_to_dd_id(trace_flags_hex)
+            # there's currently only one trace flag, which denotes sampling priority
+            # was set to keep "01" or drop "00"
+            # trace flags is a bit field: https://www.w3.org/TR/trace-context/#trace-flags
+            sampling_priority = trace_flags & 0x1
+
+        else:
+            raise ValueError("W3C traceparent hex length incorrect: %s" % tp)
+
+        return trace_id, span_id, sampling_priority
+
+    @staticmethod
+    def _get_tracestate_values(ts):
+        # type: (str) -> Optional[Tuple[Optional[int], Dict[str, str], Optional[str]]]
+
+        # tracestate parsing, example: dd=s:2;o:rum;t.dm:-4;t.usr.id:baz64,congo=t61rcWkgMzE
+        dd = None
+        ts_l = ts.split(",")
+        for list_mem in ts_l:
+            if list_mem.startswith("dd="):
+                # cut out dd= before turning into dict
+                list_mem = list_mem[3:]
+                dd = dict(item.split(":") for item in list_mem.split(";"))
+
+        # parse out values
+        if dd:
+            sampling_priority_ts = dd.get("s")
+            if sampling_priority_ts:
+                sampling_priority_ts = int(sampling_priority_ts)
+
+            origin = dd.get("o")
+            # need to convert from t. to _dd.p.
+            other_propagated_tags = {
+                re.sub("t.", "_dd.p.", k): v
+                for (k, v) in dd.items()
+                if (_TraceContext._is_valid_datadog_trace_tag_key(k))
+            }
+
+            return sampling_priority_ts, other_propagated_tags, origin
+        else:
+            log.debug("no dd list member in tracestate from incoming request: %r", ts)
+            return None
+
+    @staticmethod
+    def _get_sampling_priority(sampling_priority_tp, sampling_priority_ts):
+        # type: (int, int) -> int
+        """
+        1. If the traceparent sampled flag == 0, and the tracestate sampling priority is not present or greater than 0,
+        set the sampling priority field to 0.
+
+        2. If the traceparent sampled flag == 1, and the tracestate sampling priority is
+        not present or less than or equal to 0, update the sampling priority field to 1.
+
+        3. If the traceparent sampled flag == 1, and the tracestate sampling priority is greater than 0,
+        use the extracted sampling priority value.
+
+        4. If the traceparent sampled flag == 0, and the tracestate sampling priority is less than or equal to 0,
+        use the extracted sampling priority value.
+        """
+
+        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
```

Comment:
ah this helped me understand a typing issue I was having. My typing for _get_sampling_priority was wrong it's `# type: (Optional(int), int) -> int` rather than `# type: (int, int) -> int`

Reply:
Munir was able to simplify the logic to this though, which I've added:

```
        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
            sampling_priority = 0

        elif sampling_priority_tp == 1 and (not sampling_priority_ts or sampling_priority_ts < 0):
            sampling_priority = 1
        else:
            # elif sampling_priority_tp == 1 and sampling_priority_ts > 0:
            # elif sampling_priority_tp == 0 and sampling_priority_ts < 0:
            sampling_priority = sampling_priority_ts
```